Improve performance with /ds

SolarisFox opened the pull request:
removed unnecessary uses of ```Tools.getMove()``` and ```Tools.getTemplate()```

Average search time currently reduced to around 20ms. (excluding recovery search)

codelegend commented on config/commands.js:
```diff
@@ -676,18 +676,21 @@ var commands = exports.commands = {
 						if (!template.learnset) template = Tools.getTemplate(template.baseSpecies);
 						if (!template.learnset) continue;
 						for (var i in searches[search]) {
-							var move = Tools.getMove(i);
```
why not keep the new cache object as `moves`, and this line as `var move = moves[i];`
So, you need not change the further lines from `move` to `move[i]`.

SolarisFox replied:
Having to copy the whole move object every time is what was slowing it down so much. Sense the only property that is being used in this command is ```move.id``` I suppose I could create something like ```moveId = move[i].id;``` . I'm not sure that's any more clear then the way I have it now though.

codelegend replied:
objects in javascript dont get copied, they get referenced, so it takes literally 0 time

codelegend replied:
it was the function call `getMove` that was slowing it down.
Using a direct reference, instead of dereferencing the property each time makes it faster actually.
And yes, if only `move.id` is used, then you could just say `moveId = moves[i].id`

codelegend commented on config/commands.js:
```diff
@@ -676,18 +676,21 @@ var commands = exports.commands = {
 						var template = Tools.getTemplate(dex[mon].id);
 						if (!template.learnset) template = Tools.getTemplate(template.baseSpecies);
 						if (!template.learnset) continue;
 						for (var i in searches[search]) {
```
make `searches[search]` into `moves`?

codelegend commented on config/commands.js:
```diff
@@ -676,18 +676,21 @@ var commands = exports.commands = {
 						for (var i in searches[search]) {
-							var move = Tools.getMove(i);
-							if (!move.exists) return this.sendReplyBox("'" + move + "' is not a known move.");
 							var prevoTemp = Tools.getTemplate(template.id);
```
can't this be moved out of the `for` loop too?

codelegend replied:
you could cache all `prevoTemp`, so that you can avoid calls to `getTemplate` each time.

codelegend commented on config/commands.js:
```diff
@@ -676,18 +676,22 @@ var commands = exports.commands = {
-							if (!move.exists) return this.sendReplyBox("'" + move + "' is not a known move.");
+						for (var i in moves) {
+							var moveId = moves[i].id;
 							var prevoTemp = Tools.getTemplate(template.id);
```
why not cache all pre-evolutions? saves time on `getTemplate`, as number of pre-evolutions is at max 2

SolarisFox replied:
I'll compile a list of searches to try with the cached prevoTemps to see the actual impact that has, because it slows the search down slightly in cases of only searching one move.

monsanto commented:
If you are serious about improving the performance of this method, you are going about it the wrong way with super microoptimizations like caching the result of `move.id`. First, that function is *way* too big,

monsanto commented:
Sorry, misclick.

monsanto commented:
For posterity, here is what I had to say about this:

~chaos: if you are serious about improving the performance of dexsearch(), microoptimizations like caching move.id are the wrong way of going about it
~chaos: the biggest problem with this method that i see just from a cursory glance, is that you are using for-in
~chaos: there are roughly two general ways of using an object in javascript, one is like a C struct, and the other is like a hash table
~chaos: PS uses object as hash table extensively
~chaos: the problem is that if you use for-in on an object in hash table mode, you will disable any optimizations to the function
~chaos: and since that function is huge as fuck, you are paying a huge penalty for the deopt
~chaos: the easiest way to improve the speed of that function is to change every instance of for-in to a for(;;) loop with Object.keys
~chaos: but even then there may be other things in that function that are disabling optimization
~chaos: so the next step would be to break up the function a bit so you dont pay such a huge price for a deopt
~chaos: (i am not sure how familiar you are with v8 performance tuning)
+SolarisFox: Not very. I at least understand the things you're saying to change though.
~chaos: https://github.com/petkaantonov/bluebird/wiki/Optimization-killers this is a good reference
+codelegend: https://github.com/Zarel/Pokemon-Showdown/blob/master/config/commands.js#L625
+codelegend: this could be made into a if-else
+codelegend: removes the for loop and the switch?
~chaos: that line is really odd
~chaos: even if you did not want to change it to an if statement, it will be *much* faster to loop over an array of ['gen', 'tier', ...]
+SolarisFox: Yeah, I think TTT did it so he could experiment with the effect of re-ordering the when they were searched without moving blocks of code
+SolarisFox: (that and he just loves switch statements for some reason)
~chaos: (one could make the argument that v8 should be smart enough to recognize the object is a constant, and thus the for loop is not actually a loop and can be unrolled, BUT...
~chaos: ... you are completely disabling the optimizer by using a for-in anyway)

Zarel commented:
@monsanto, we support `/eval` so I'd be surprised if any of our functions are optimizable.

Slayer95 commented:
<code>eval()</code> is called within functions, so that's not a problem (except for the associated command and the repl.

monsanto commented:
@Zarel `eval` only deopts the function it is called within. It doesn't make sense for `eval` to corrupt other functions, because anything that `eval` could do to corrupt an optimized function, a function that doesn't use eval could do as well.

monsanto commented:
There are still a couple of for-ins. Note that a for-in on an array will always result in a deopt. I didn't see them last time or I would have pointed it out. The only case where for-in does not result in a deopt is a "non-array" object with no enumerable properties on the prototype stored in a local variable (lol). From what I know about Google's plans for V8, my guess is this isn't changing any time soon (their "strong mode" proposal makes for-in a syntax error). 

If you are writing fast JavaScript you should just never use for-in, it is too finicky, and `Object.keys` covers all cases except searching prototypes for enumerable properties.

monsanto commented:
Are there any cases where you access an element of `dex`, where the key does not come from a loop over `Object.keys`? That is, are you actually making use of the hash table you have allocated? It doesn't seem so to me. Using an array will be much faster.

Zarel commented:
Have we ever `delete`d from or assigned to `Tools.data.Pokedex`? In theory the answer is no.

monsanto commented:
My comment on SolarisFox's commit about delete seems to have disappeared when they updated the branch. I must have fucked up something with GH because I don't get a cool "commented on an outdated diff" message... anyway, since this information may be useful for someone in the future, here is comment:

> `delete`ing a property from an object (incl. arrays) forces the internal representation of the object to be a hash table. This means roughly three things: 1) your entire array will be copied, which is slow; 2) any indexing will no longer use a fast load offset instruction, but instead a slow hash table access (think Python speed), 3) any inline caches that are expecting `dexKeys` to be an array will be invalidated, and they will transition from an efficient monomorphic state to a slower polymorphic state. If you are unfamiliar with that terminology and are interested in writing fast JavaScript, [read this post](http://mrale.ph/blog/2015/01/11/whats-up-with-monomorphism.html).

SolarisFox commented:
moves, abilities, tiers, and type run at about 20ms
stat ranges run at about 11ms
colors and gen run at about 18ms
recovery takes around 68ms (10 moves)

if everything ~~looks good~~ functions as expected I'll do something similar with /movesearch and rebase it onto this pullreq.

Zarel commented:
Well, as long as it's an improvement, it's fine.

@SolarisFox, can you benchmark whether or not `Object.keys` is actually an improvement?

monsanto commented:
I am benchmarking it, since I need to check V8s trace for other deoptimizations anyway.

xfix commented:
@monsanto

> It doesn't make sense for eval to corrupt other functions

```javascript
var scope = function scope() {
    var someScopedVariable = 'Here you go.'
    function runCode(code) {
        return eval(code)
    }

    return {runCode: runCode}
}()

alert(scope.runCode('someScopedVariable'))
```

A variable in this case cannot be removed after leaving a scope, because `eval` possibly may refer to it. `eval` affects the entire scope and all scopes above it, because it can suddenly change values of variables in unpredictable way (imagine running `someScopedVariable = 'hahaha changed'`, which prevents optimizations like constant folding.

Generally speaking, if you don't want this performance penalty, use `new Function(code)()` or [indirect eval](https://es5.github.io/#x10.4.2). It's still somewhat slow, because it runs JavaScript parser, but at least it doesn't affect code outside. However, this only allows accessing global scope, which may or may not be a good thing (`new Function` can somewhat deal with this limitation by using function parameters).

Slayer95 commented:
Why wouldn't a good compiler just remove every optimization when <code>eval</code> is run? If <code>eval</code> wrecks that much, then it looks like V8 is currently lacking. We shouldn't write code designed to run specifically by a certain version of a JIT compiler. Note that JS server runtimes aren't even intrinsecally limited to using <code>V8</code>, and there is at least one Node fork (JXCode) which uses Spidermonkey and therefore supports optimizing the elusive <code>try..catch</code> blocks.

I would even argue that, by principle, developer-made optimizations should be designed for interpreted code. However, there is a pragmatic argument against which this can't stand.

monsanto commented:
@xfix you are right, I was being imprecise; direct eval does at the moment deoptimize enclosing functions. This case doesn't concern PS though. I'm not sure why you think this has anything to do with constant folding, because an inner function that does `someScopedVariable = unpredictableThing() ? 1 : 0` will disable constant folding too.

monsanto commented:
@Slayer95 It doesn't wreck anything, what @xfix is talking about is forcing local variables to be context allocated (like global variables). Currently V8 deoptimizes functions with a context-allocated stack, but functions that simply *use* these variables are fine (if this wasn't the case, every function would be deoptimized if it referenced a global variable). You can see this behavior by doing the following:

```
var scope = function scope(x) {
    "use strict";
    var someScopedVariable = 'Here you go.'
    
    function runCode(code) {
        eval(code)
    }

    function ayy(code) {
       console.log(someScopedVariable)
    }   

    return {runCode: runCode, ayy: ayy}
}();

%OptimizeFunctionOnNextCall(scope.runCode);
%OptimizeFunctionOnNextCall(scope.ayy);
scope.runCode("someScopedVariable = 'changed'");
scope.ayy();
console.log(%GetOptimizationStatus(scope.runCode)); // 2 (unoptimized)
console.log(%GetOptimizationStatus(scope.ayy)); // 1 (optimized)
```

Slayer95 commented:
Oh, I see... No problem then.

monsanto commented:
OK, I am done benchmarking. This is on doing `/dexsearch meteor mash`:

```
no changes:
node 0.10, no changes 15ms-20ms
io.js, no changes 9ms-12ms

changes:
node 0.10, changes 8-11ms (I cannot get the optimizer to run ?_? whatever we are switching to io.js anyway)
io.js, 2-4ms.
```

There are still a few bits of low hanging fruit but I think at this point I am going to bow out--this isn't a critical function, and I've had my fun.

@Zarel in this case there is no benefit to using `Object.keys` over `for-in` because--to my surprise--the objects in `Tools.data.Pokedex` and friends aren't actually being stored as hash tables. This is because the data specified as a literal and ran through (the equivalent of) `eval`. One of the implications of this is that if you change this data, and hotpatch the server, you'll thrash V8's inline caches a bit, but it probably doesn't matter. If you care, you can force it to be a hash table by `delete`ing a property, using `JSON.parse` instead of `require`, or using a `Map`.

I am gathering you are asking the question because you want to know if it's a big enough deal to use `Object.keys` everywhere. If that's the case, I think that is missing the point a bit; `Object.keys` runs at roughly the same speed as `for-in`. The real issue is whether or not optimization matters for the function in question, which who knows ¯\\\_(ツ)\_/¯ the optimizer is also less effective if other functions that are inlining candidates can't be optimized, so you may not even see any benefit from optimization until you clean up everything else. If you want to experiment with it, I am guessing almost every function in rooms.js is prevented from being optimized/inlined by the `for-in` loops over `this.users`, `this.bannedUsers`, and `this.bannedIps`.

I hope people don't take this topic as some suggestion that I think everything should be optimized, because there are much more pressing things to do. I was just bored and was annoyed with bad performance advice.

monsanto commented:
For anyone interested in the performance of `for-in`, it is suggested that TurboFan (new optimizing compiler in development, announced August 2014) no longer deopts. Crankshaft (the current optimizing compiler) will never get full support. https://code.google.com/p/v8/issues/detail?id=3913#c3

Here's to them fixing `try-catch` deopts sometime soon.

SolarisFox commented:
Because for-in seems to not be any slower, and the other changes by themselves make a substantial difference I'll leave the loops how they were previously. If somebody who understands understands the deopts better then I do wants to restructure the command I'll leave that for them.

SolarisFox commented:
Also made two very minor changes to /movesearch to match structure changes in /dexsearch.

SolarisFox commented:
changed to ```typeof searches[search][stat].less === 'number'``` to stat range searches to prevent /ds atk = 0 from returning all pokemon when it should return none.

Also changed proper range searches in /movesearch to match the structure of /dexsearch

Zarel commented:
> @Zarel in this case there is no benefit to using `Object.keys` over `for-in` because--to my surprise--the objects in `Tools.data.Pokedex` and friends aren't actually being stored as hash tables. This is because the data specified as a literal and ran through (the equivalent of) `eval`. One of the implications of this is that if you change this data, and hotpatch the server, you'll thrash V8's inline caches a bit, but it probably doesn't matter. If you care, you can force it to be a hash table by `delete`ing a property, using `JSON.parse` instead of `require`, or using a `Map`.

I mean, it's basically a read-only database; of course we'd never add or delete entries.

Would lookups be faster as a hash table?
